### PR TITLE
fix: deal with new lints from beta clippy

### PR DESCRIPTION
**Pull request by thomaseizinger**
## Description

Most of this is trivial, apart from the rename of the `clippy::derive_hash_xor_eq` lint to `clippy::derived_hash_with_manual_eq`.

Instead of allowing that lint, we manually implement `PartialEq` and add a comment why the difference between the `PartialEq` and `Hash` implementations are okay.

## Change checklist

- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] A changelog entry has been made in the appropriate crates

### CODE REVIEW

**Comment by elenaf9** [comment]
I would prefer if we simply resolve the lint, i.e. manually implement `PartialEq` for `LocalStreamId` so that it respects the (imo reasonable) invariant `v1 == v2 <=> h(v1) == h(v2)`:

```rust 
impl PartialEq for LocalStreamId {
    fn eq(&self, other: &Self) -> bool {
        self.num == other.num
    }
}
```

Else if `PartialEq` should also compare the `role` field we can add that in the manual implementation, but then I think it deserves a comment stating why it is okay to do this (I assume it is because both sides agree on the role for each stream).

File: muxers/mplex/src/codec.rs [hunk 1]
```diff
@@ -65,7 +65,12 @@ impl fmt::Display for LocalStreamId {
 }
 
 impl Hash for LocalStreamId {
-    #![allow(clippy::derive_hash_xor_eq)]
+    #![allow(
+        unknown_lints, // Prior to Rust 1.68, ignore the new `derived_hash_with_manual_eq` lint.
+        renamed_and_removed_lints, // After Rust 1.68, ignore usage of the old `derive_hash_xor_eq` lint.
+        clippy::derive_hash_xor_eq, // Prior to Rust 1.68, silence the lint.
+        clippy::derived_hash_with_manual_eq // After Rust 1.68, silence the lint.
```
Comment (elenaf9) [inline muxers/mplex/src/codec.rs:72]: That lint name is confusing. We are actually doing it the other way around here: we derive eq and manually implement hash.

**Comment by thomaseizinger** [comment]
Of course, that is a much simpler way forward. Thanks @elenaf9!

**Review by elenaf9 — APPROVED** [review]

**Review by jxs — APPROVED** [review]

**Comment by thomaseizinger** [comment]
@elenaf9 @jxs So I actually had to replicate the exact implementation of the derived `PartialEq` to make the tests pass again. For some reason, this even affected network compatibility (interop tests wuhu!).

**Comment by thomaseizinger** [comment]
@mergifyio backport rust-v050-testplan

**Comment by mxinden** [comment]
> @ Mergifyio backport rust-v050-testplan

:exploding_head:

**Comment by thomaseizinger** [comment]
> > @ Mergifyio backport rust-v050-testplan
> 
> :exploding_head: 

😉

**Pull request merged**